src/lib/edit_script.rs

```diff
@@ -0,0 +1,330 @@
+                                                       x: NodeId,
+                                                       script: &mut EditScript<T>,
+                                                       new_mappings: &TemporaryMappingStore,
+                                                       from_in_order: &mut HashSet<NodeId>,
+                                                       to_in_order: &mut HashSet<NodeId>)
+                                                       -> Result<(), ArenaError> {
+        debug!("align_children({}, {})", w, x);
+        for child in x.children(&store.to_arena) {
+            to_in_order.remove(&child);
+        }
+        let mut s1: Vec<NodeId> = vec![];
+        for child in w.children(&store.from_arena) {
+            from_in_order.remove(&child);
+            if new_mappings.contains_from(&child) {
+                let mapped = new_mappings.get_to(&child).unwrap();
+                if x.children(&store.to_arena).any(|n| n == mapped) {
+                    s1.push(child);
+                }
+            }
+        }
+        let s2: Vec<NodeId> = vec![];
+        for child in x.children(&store.to_arena) {
+            if new_mappings.contains_to(&child) {
+                let mapped = new_mappings.get_from(&child).unwrap();
+                if w.children(&store.from_arena).any(|n| n == mapped) {
+                    s1.push(child);
+                }
+            }
+        }
+        let lcs = self.lcss(store, &s1, &s2);
+        for &(from, to) in &lcs {
+            from_in_order.insert(from);
+            to_in_order.insert(to);
+        }
+        for a in &s1 {
+            for b in &s2 {
+                if new_mappings.contains_from(a) && new_mappings.get_to(a).unwrap() == *b &&
+                   !lcs.contains(&(*a, *b)) {
+                    let k = self.find_pos(store, *b, new_mappings, to_in_order);
+                    let mut mov = Move::new(*a, w, k);
+                    debug!("Edit script: MOV {} {:?} Parent: {} {:?}",
+                           store.from_arena[*a].label,
+                           store.from_arena[*a].value,
+                           store.to_arena[w].label,
+                           store.to_arena[w].value);
+                    script.push(mov);
+                    mov.apply(&mut store.from_arena)?;
+                    from_in_order.insert(*a);
+                    to_in_order.insert(*b);
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Find the position of node x in to_arena.
+    fn find_pos<T: Clone + Debug + Eq + 'static>(&self,
+                                                 store: &MappingStore<T>,
+                                                 x: NodeId,
+                                                 new_mappings: &TemporaryMappingStore,
+                                                 to_in_order: &HashSet<NodeId>)
+                                                 -> u16 {
```

Comment:
`u16` seems like quite a small type. Is that deliberate?

Reply:
This function finds out where a node is in a list of children (i.e. "this is the third child of its parent"), and it seemed unlikely that the number of children would exceed the size of a `u16`, but maybe I'm naive about grammars?

Reply:
Ah, so it's only the direct children? I guess it's OK. But do you ever store this `u16` anywhere? I'm fine with taking a chance if we can cut storage space down, but if it's just a parameter we pass between functions, `u32` might be just as performant in practise? It might be worth knowing that in release mode, rustc's checks on overflow and the like are off.

Reply:
If we change this we also need to change some code in `action.rs` which is not changed in this PR (this PR doesn't change any functionality). I'll raise an issue for it and we can think about it another time.

Reply:
Good idea.